---
title: "Quality check, differential expression and splicing analysis"

output:
  pdf_document:
    toc: yes
    toc_float: no
---

# Experiment info

```{r}
# metadata.txt is a tab separated file with the info of samples and conditions. It is the same file used a sample metadata for splicing analysis in the deseq2_dexseq_isoformswitchanalyzer.Rscript
library(DT)
library(reshape2)
library(dplyr)
sample_info <- read.table("metadata.txt",header = TRUE,sep = "\t")
DT::datatable(sample_info)
```

# Preprocessing statistics

## Alignment to genome - STAR

```{r}
library(reshape2)
library(dplyr)
library(ggplot2)
library(viridis)
# Import STAR alignment statistics
alignment_stats_star <- read.table(file = "multiqc_data/multiqc_star.txt",header = TRUE,sep = "\t")
alignment_stats_star <- melt(data = alignment_stats_star,id.vars="Sample")#,"total_reads","avg_input_read_length"))

#Plot percentage of reads aligned to the genome
alignment_stats_star$value <- as.numeric(alignment_stats_star$value)
renderPlot({
alignment_stats_star %>%
  dplyr::filter(grepl(pattern = "percent",x = alignment_stats_star$variable)) %>%
  ggplot(aes(x = value, y = Sample)) +
  geom_bar(aes(fill = variable), stat="identity") + 
  scale_fill_viridis(discrete = TRUE,alpha = 0.75,option = "D",direction = 1) +
  theme_classic()+
  ggtitle(label = "Percentage of reads aligned with STAR")+
  theme(#legend.position=c(0.1, 0.7),
        axis.text.x = element_text(angle = 45,hjust = 1),
        axis.ticks = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_rect(fill = "transparent",colour = NA),
        plot.background = element_rect(fill = "transparent",colour = NA),
        legend.background = element_rect(fill = "transparent", colour = NA),
        legend.box.background = element_rect(fill = "transparent", colour = NA))
})
```


```{r}
renderPlot({
alignment_stats_star %>%
  dplyr::filter(grepl(pattern = "read_length",x = alignment_stats_star$variable)) %>%
  ggplot(aes(x = value, y = variable)) +
  geom_bar(aes(fill=Sample),position = "dodge",stat="identity") + 
  scale_fill_viridis(discrete = TRUE,alpha = 0.75,option = "D",direction = 1) +
  theme_classic()+
  ggtitle(label = "Percentage of reads aligned with STAR")+
  theme(#legend.position=c(0.1, 0.7),
        axis.text.x = element_text(angle = 45,hjust = 1),
        axis.ticks = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_rect(fill = "transparent",colour = NA),
        plot.background = element_rect(fill = "transparent",colour = NA),
        legend.background = element_rect(fill = "transparent", colour = NA),
        legend.box.background = element_rect(fill = "transparent", colour = NA))
})
```


## Pseudoalignment to transcriptome - Kallisto

```{r, echo=FALSE, warning=FALSE, message=FALSE,fig.height=5,fig.width=10,dev.args = list(bg = 'transparent')}
library(ggplot2)
library(reshape2)
library(dplyr)
library(plotly)
alignment_stats_kallisto <- read.csv("multiqc_data/multiqc_kallisto.txt",header = TRUE,sep = "\t")
alignment_stats_kallisto <- melt(data = alignment_stats_kallisto,id.vars="Sample")#,"total_reads","avg_input_read_length"))
renderPlot({
alignment_stats_kallisto %>%
  dplyr::filter(grepl(pattern = "percent",x = alignment_stats_kallisto$variable)) %>%
  ggplot(aes(x = value, y = Sample)) +
  geom_bar(aes(fill = variable), stat="identity",) +
  scale_fill_viridis(discrete = TRUE,alpha = 0.75,option = "D",direction = 1)+
  xlim(0,100)+
  theme_classic()+
  ggtitle(label = "Percentage of reads pseudoaligned with Kallisto")+
  theme(#legend.position=c(0.1, 0.7),
        axis.text.x = element_text(angle = 45,hjust = 1),
        axis.ticks = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_rect(fill = "transparent",colour = NA),
        plot.background = element_rect(fill = "transparent",colour = NA),
        legend.background = element_rect(fill = "transparent", colour = NA),
        legend.box.background = element_rect(fill = "transparent", colour = NA))
})
```

# Verification of biological conditions

## Distance between samples 

```{r}
library(RColorBrewer)
library(pheatmap)
library(DESeq2)
library(ggpubr)
load(file = "gene_counts/dds.RData")
vst <- vst(dds, blind=FALSE)
sampleDists <- dist(t(assay(vst)))
sampleDistsMatrix <- as.matrix(sampleDists)

colors <- colorRampPalette( rev(brewer.pal(length(dds$sample), "Blues")) )(255)

renderPlot({
pheatmap(sampleDistsMatrix,clustering_distance_rows=sampleDists,clustering_distance_cols=sampleDists,col=colors,legend = FALSE,main = "Sample distances",treeheight_row = 10,treeheight_col = 20)
})
```

## Sanity check of the datasets | PCA of samples

```{r}
PCA_plot <- plotPCA(vst, intgroup=c("condition"))
renderPlot({
PCA_plot +
    theme_classic(base_size = 12)+
    theme(#legend.position=c(0.1, 0.9),
          aspect.ratio = 1,
          plot.title = element_text(size = 12),#color = "red",, face = "bold"),
          axis.text.x = element_text(size = 8),
          axis.title.x = element_text(size = 10),
          axis.text.y = element_text(size = 8),
          axis.title.y = element_text(size = 10),
          legend.position = "left",
          axis.ticks = element_blank(),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_rect(fill = "transparent",colour = NA),
          plot.background = element_rect(fill = "transparent",colour = NA),
          legend.background = element_rect(fill = "transparent", colour = NA),
          legend.box.background = element_rect(fill = "transparent", colour = NA))
})
```

## Sanity check of the datasets | Expression of specific genes

```{r}

library(DESeq2)
library(ggplot2)
degs<-read.table(file = "gene_counts/degs_deseq2.txt",header = TRUE,sep = "\t")

selectInput("gene_name", label = "Gene",
              choices = unique(degs$ensembl_gene_id_version))

renderPlot({
  gene_counts <- plotCounts(dds, 
                           gene=input$gene_name, 
                           intgroup="condition", 
                           returnData=TRUE)
  
ggplot(gene_counts, aes(x = condition, y = count,col=condition)) +
  geom_point() +
  scale_color_brewer(palette = "Set1",)+
    theme_classic(base_size = 12)+
    # ggtitle(label = i,subtitle = element_blank())+
    theme(#legend.position=c(0.9, 0.9),
          # aspect.ratio = 1,
          # axis.text.x = element_text(size = 8),
          # axis.title.x = element_text(size = 10),
          # axis.text.y = element_text(size = 8),
          # axis.title.y = element_text(size = 10),
          legend.position = "none",
          axis.ticks = element_blank(),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_rect(fill = "transparent",colour = NA),
          plot.background = element_rect(fill = "transparent",colour = NA),
          legend.background = element_rect(fill = "transparent", colour = NA),
          legend.box.background = element_rect(fill = "transparent", colour = NA))+
  scale_y_log10()+
  labs(y="log10(normalized_counts)",col="Condition")
})
```

## Sanity check - Volcano plot

```{r}
library(dplyr)
library(biomaRt)
library(ggplot2)
degs<-read.table(file = "gene_counts/degs_deseq2.txt",header = TRUE,sep = "\t")
degs$ensembl_gene_id <- gsub(pattern = "\\..*$",replacement = "",x=degs$ensembl_gene_id_version)

mart_export <- read.delim(file = "databases/mart_export.txt")

degs <- dplyr::left_join(degs,mart_export,by="ensembl_gene_id")
renderPlot({
ggplot(data=degs,aes(x=log2FoldChange,y=-log10(padj))) +
  geom_vline(xintercept=c(-log2(1.5),log2(1.5)), color="red")+ 
  geom_hline(yintercept=-log10(0.05), color="blue")+ 
  geom_point(color="black",alpha=0.5,stat="identity")+
  xlim(-10,10)
  })
```

# Differential gene expression 

## Expression of splicing factors

```{r}
sf_genes<- read.delim(file = "databases/splicing_factors_list.txt")

library(ggplot2)
library(RColorBrewer)
heatmap_sf <- degs[degs$external_gene_name %in% sf_genes$Gene,]
heatmap_sf$padj[is.na(heatmap_sf$padj)] <- 1
renderPlot({
ggplot(data=heatmap_sf,aes(x=log2FoldChange, y=external_gene_name, size = 1-padj,color=log2FoldChange)) +
  geom_point() +
  # xlim(-1,1)+
  # scale_size(range = c(1, 10), name="padj")+
  theme_classic()+
  geom_vline(xintercept=c(-log2(1.5),log2(1.5)), color="black")+
  scale_color_gradient2(midpoint=0, low="#313695", mid="#FFFFBF",
                     high="#A50026", space ="Lab" )
})
```

## The list of DEGs

```{r echo=FALSE,warning=FALSE,message=FALSE}
library(DT)
library(dplyr)
degs %>%
  dplyr::select(external_gene_name,ensembl_gene_id,log2FoldChange,padj,baseMean,lfcSE,stat,pvalue) %>%
  datatable()
```


# Splicing analysis - Isoform-level (IsoformSwitchAnalyzer)

## Isoform-level (IsoformSwitchAnalyzer) - number of genes

```{r}
library(IsoformSwitchAnalyzeR)
load(file = "pseudocounts/exampleSwitchListAnalyzed.RData")
datatable(extractSwitchSummary(exampleSwitchListAnalyzed))
```

## Switchplot 
```{r,fig.height=20}
# Runs only in shiny
isa_genes <- read.delim(file = "pseudocounts/isoform_signifcant_genes.txt")

selectInput("gene_name", label = "Gene",
              choices = unique(isa_genes$gene_name))
renderPlot({
  switchPlot(exampleSwitchListAnalyzed,
             gene=input$gene_name)
})
```

## Plot Splicing summary 
```{r}
renderPlot({
extractSplicingSummary(
    exampleSwitchListAnalyzed,
    asFractionTotal = FALSE,
    plotGenes=FALSE
)
})
```

# Splicing enrichment
```{r}
renderPlot({
splicingEnrichment <- extractSplicingEnrichment(
    exampleSwitchListAnalyzed,
    splicingToAnalyze='all',
    returnResult=TRUE,
    returnSummary=FALSE
)
})
```

# Splicing genome-wide
```{r}
renderPlot({
extractSplicingGenomeWide(
    exampleSwitchListAnalyzed,
    featureToExtract = 'all',                 # all isoforms stored in the switchAnalyzeRlist
    splicingToAnalyze = c('A3','MES','ATSS'), # Splice types significantly enriched in COAD
    plot=TRUE,
    returnResult=FALSE  # Preventing the summary statistics to be returned as a data.frame
)
})
```


# Exon-level - DEXSeq 

## Number of exons and genes

```{r }
# pvalue_threshold= 0.1 ##set this parameter
# foldchange_threshold=1.5 ##set this parameter
# dexseq_genes <- read.csv(file = "exon_counts/dexs_dexseq.csv",header = TRUE,row.names = NULL)
load(file = "exon_counts/dxr1.RData")
significant_exons <- as.data.frame(dxr1) %>%
  dplyr::filter(padj < 0.1 )
genes <- unique(significant_exons$groupID)
#Number of significant exons 
print(paste0("Total number of exons evaluated: ",sum(table ( dxr1$padj < 0.1 )) ))
print(paste0("Total number of exons differentially expressed: ",table ( dxr1$padj < 0.1 )[2]) )

# Number of genes
print(paste0("Total number of genes evaluated: ",sum(table(tapply( dxr1$padj < 0.1, dxr1$groupID, any)))))
print(paste0("Total number of genes with at least one exon differentially expressed: ",table(tapply(dxr1$padj < 0.1, dxr1$groupID, any))[2]))
```


## Visualization 
```{r}
library(DEXSeq)
selectInput("gene_name", label = "Gene",
              choices = genes)
renderPlot({
  plotDEXSeq(dxr1, gene=input$gene_name, legend=TRUE, cex.axis=1.2, cex=1.3, lwd=2 )
})
```

## The full list of genes with differentially expressed exons

```{r}
library(DT)
DT::datatable(significant_exons,class="display",rownames=F,filter="top")#,editable=F,extensions = "Buttons",options = list(dom="Bfrtip",buttons=c("copy","csv","excel","pdf","print")))
```

# Event-level - MAJIQ 

## Classify events by event type

```{r}
## Import original majiq tsv output files and parse them to generate the dataframes containing positive  and negative events for each event type

#voila_res_all --> dataframe with original tsv output file from majiq and voila with ALL events detected
#voila_res --> dataframe with original tsv output file from majiq with POSITIVE events only
"%notin%" <- Negate("%in%")
library(dplyr)

voila_res_all <- read.delim(file = "majiq_output/voila_results_all.tsv",header = TRUE,sep = "\t")
voila_res <- read.delim(file = "majiq_output/voila_results.tsv",header = TRUE,sep = "\t")

################
## Exon Skipping
################

# majiq.es.all --> df with ALL ES events detected
# majiq.es.pos --> df with POSITIVE ES events
# majiq.es.neg --> df with NEGATIVE ES events

  ## Get all ES events
majiq.es.all <- voila_res_all %>%
  dplyr::filter(ES =="True" & A5SS =="False" & A3SS =="False" & IR.coords=="")  %>%
  dplyr::filter(!grepl('na', LSV.ID))

  ## Get POSITIVE ES events
majiq.es.pos <- voila_res %>%
  dplyr::filter(ES =="True" & A5SS =="False" & A3SS =="False" & IR.coords=="")  %>%
  dplyr::filter(!grepl('na', LSV.ID))

  ## Get NEGATIVE ES events
temp <- data.frame(do.call("rbind", strsplit(x = majiq.es.all$E.dPSI..per.LSV.junction,split = ";",fixed = TRUE)))
temp  <- as.data.frame(apply(temp, 2, as.numeric))
keep <- apply(temp < 0.001 & temp > -0.001, 1, all, na.rm=TRUE)
temp  <- as.data.frame(cbind(temp,majiq.es.all$LSV.ID))
majiq.es.neg <- majiq.es.all[keep,]
rm(temp,keep)

################
## A3SS
################

# majiq.a3.all --> df with ALL A3 events
# majiq.a3.pos --> df with POSITIVE A3 events
# majiq.a3.neg --> df with NEGATIVE A3 events

  ## Get all A3 events
majiq.a3.all <- voila_res_all %>%
  dplyr::filter(ES =="False" & A5SS =="False" & A3SS =="True" & IR.coords=="")  %>%
  dplyr::filter(!grepl('na', LSV.ID))

  ## Get POSITIVE A3 events
majiq.a3.pos <- voila_res %>%
  dplyr::filter(ES =="False" & A5SS =="False" & A3SS =="True" & IR.coords=="")  %>%
  dplyr::filter(!grepl('na', LSV.ID))

  ## Get NEGATIVE A3 events
temp <- data.frame(do.call("rbind", strsplit(x = majiq.a3.all$E.dPSI..per.LSV.junction,split = ";",fixed = TRUE)))
temp  <- as.data.frame(apply(temp, 2, as.numeric))
keep <- apply(temp < 0.001 & temp > -0.001, 1, all, na.rm=TRUE)
temp  <- as.data.frame(cbind(temp,majiq.a3.all$LSV.ID))
majiq.a3.neg <- majiq.a3.all[keep,]
rm(temp,keep)

################
## A5SS
################

# majiq.a5.all --> df with ALL A5 events
# majiq.a5.pos --> df with POSITIVE A5 events
# majiq.a5.neg --> df with NEGATIVE A5 events

  ## Get all A5 events
majiq.a5.all <- voila_res_all %>%
  dplyr::filter(ES =="False" & A5SS =="True" & A3SS =="False" & IR.coords=="")  %>%
  dplyr::filter(!grepl('na', LSV.ID))

  ## Get POSITIVE A5 events
majiq.a5.pos <- voila_res %>%
  dplyr::filter(ES =="False" & A5SS =="True" & A3SS =="False" & IR.coords=="")  %>%
  dplyr::filter(!grepl('na', LSV.ID))

  ## Get NEGATIVE A5 events
temp <- data.frame(do.call("rbind", strsplit(x = majiq.a5.all$E.dPSI..per.LSV.junction,split = ";",fixed = TRUE)))
temp  <- as.data.frame(apply(temp, 2, as.numeric))
keep <- apply(temp < 0.001 & temp > -0.001, 1, all, na.rm=TRUE)
temp  <- as.data.frame(cbind(temp,majiq.a5.all$LSV.ID))
majiq.a5.neg <- majiq.a5.all[keep,]
rm(temp,keep)

################
## IR
################

# majiq.ir.all --> df with ALL IR events
# majiq.ir.pos --> df with POSITIVE IR events
# majiq.ir.neg --> df with NEGATIVE IR events

  ## Get all IR events
majiq.ir.all <- voila_res_all %>%
  dplyr::filter(ES =="False" & A5SS =="False" & A3SS =="False" & IR.coords!="")  %>%
  dplyr::filter(!grepl('na', LSV.ID))

  ## Get POSITIVE IR events
majiq.ir.pos <- voila_res %>%
  dplyr::filter(ES =="False" & A5SS =="False" & A3SS =="True" & IR.coords!="")  %>%
  dplyr::filter(!grepl('na', LSV.ID))

  ## Get NEGATIVE IR events
temp <- data.frame(do.call("rbind", strsplit(x = majiq.ir.all$E.dPSI..per.LSV.junction,split = ";",fixed = TRUE)))
temp  <- as.data.frame(apply(temp, 2, as.numeric))
keep <- apply(temp < 0.001 & temp > -0.001, 1, all, na.rm=TRUE)
temp  <- as.data.frame(cbind(temp,majiq.ir.all$LSV.ID))
majiq.ir.neg <- majiq.ir.all[keep,]
rm(temp,keep)

################
## Spliced genes with more than one event happening
################
majiq.moreThanOneEvent.pos <- voila_res %>%
  dplyr::filter(LSV.ID %notin% c(majiq.es.pos$LSV.ID,majiq.a3.pos$LSV.ID,majiq.a5.pos$LSV.ID,majiq.ir.pos$LSV.ID))
```

## Number of events

```{r}
majiq.NrGenes <- data.frame(rbind(
  c("ES",nrow(majiq.es.pos)),
  c("IR",nrow(majiq.ir.pos)),
  c("A3SS",nrow(majiq.a3.pos)),
  c("A5SS",nrow(majiq.a5.pos)),
  c("Complex",nrow(majiq.moreThanOneEvent.pos))
))
colnames(majiq.NrGenes) <- c("Event type","Number of genes")
majiq.NrGenes$`Number of genes` <- as.numeric(majiq.NrGenes$`Number of genes`)
majiq.NrGenes$fraction <- majiq.NrGenes$`Number of genes` / sum(majiq.NrGenes$`Number of genes`)
majiq.NrGenes$ymax <- cumsum(majiq.NrGenes$fraction)
majiq.NrGenes$ymin <- c(0, head(majiq.NrGenes$ymax, n=-1))
majiq.NrGenes$labelPosition <- (majiq.NrGenes$ymax + majiq.NrGenes$ymin) / 2
majiq.NrGenes$label <- paste0(majiq.NrGenes$`Event type`, "\n", majiq.NrGenes$`Number of genes`)

renderPlot({
ggplot(majiq.NrGenes, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=`Event type`)) +
  geom_rect() +
  geom_label( x=c(3.5,3.0,3.5,4,3), aes(y=labelPosition, label=label), size=5) +
  scale_fill_brewer(palette=4) +
  coord_polar(theta="y") +
  xlim(c(2, 4)) +
  theme_void() +
  theme(legend.position = "none")
})
```

```{r}
majiq.heatmap <- majiq.moreThanOneEvent.pos %>% 
  dplyr::select(A5SS,A3SS,ES,IR.coords) %>%
  mutate(IR.coords=if_else(IR.coords!="", true="True",false="False"))

majiq.heatmap <- as.data.frame(table(majiq.heatmap))

majiq.heatmap <- majiq.heatmap %>%
  dplyr::filter(!(A3SS=="True" & A5SS=="False" & IR.coords=="False" & ES=="False"))%>%
  dplyr::filter(!(A3SS=="False" & A5SS=="True" & IR.coords=="False" & ES=="False"))%>%
  dplyr::filter(!(A3SS=="False" & A5SS=="False" & IR.coords=="True" & ES=="False"))%>%
  dplyr::filter(!(A3SS=="False" & A5SS=="False" & IR.coords=="False" & ES=="True"))%>%
  dplyr::filter(!(A3SS=="False" & A5SS=="False" & IR.coords=="False" & ES=="False"))
  
majiq.heatmap <- t(majiq.heatmap)
rownames(majiq.heatmap) <- c("A5SS","A3SS","ES","IR","Freq")
library(ComplexHeatmap)
col_ha = HeatmapAnnotation(NrGenes = anno_barplot(as.numeric(majiq.heatmap[5,]),
                                                  add_numbers=TRUE,
                                                  height = unit(7,"cm"),gp=gpar(fill=RColorBrewer::brewer.pal(n = 3,name = "Dark2")[3])),
                           Genes = anno_simple(as.numeric(majiq.heatmap[5,]),pch = majiq.heatmap[5,]))
renderPlot({
Heatmap(matrix = majiq.heatmap[1:4,],
        col = c("False"="White","True"="grey"),#RColorBrewer::brewer.pal(n = 3,name = "Dark2")[3]),
        border = TRUE,
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        height = unit(3,"cm"),
        name = "mat", 
        show_heatmap_legend = FALSE,
        width = unit(11,"cm"),
        top_annotation = col_ha,
        cell_fun = function(j, i, x, y, width, height, fill) {
        grid.rect(x = x, y = y, width = width, height = height, 
            gp = gpar(col = "black", fill = NA))})
})
```

# Comparative analysis of the different splicing tools' results

```{r}
genes.iso <- read.delim(file = "pseudocounts/isoform_signifcant_genes.txt")
genes.iso <- unique(genes.iso$gene_id)

load(file = "exon_counts/dxr1.RData")
genes.exo <- as.data.frame(dxr1) %>%
  dplyr::filter(padj < 0.1 )
genes.exo <- unique(genes.exo$groupID)

genes.majiq <- read.delim(file = "majiq_output/voila_results.tsv",header = TRUE,sep = "\t")
genes.majiq <- unique(genes.majiq$Gene.ID)

library(VennDiagram)
venn.splicing <- venn.diagram(x = list(DEXseq=genes.exo,
                      IsoformSwitchAnalyzer=genes.iso,
                      Majiq=genes.majiq),
             filename = NULL,
             col="black",
             fill=RColorBrewer::brewer.pal(n = 3,name = "Dark2")
)
renderPlot({
grid.newpage(); grid::grid.draw(venn.splicing)
})
```

# ChIP-seq data

## Variables

```{r}
user_chip_name <- "YBX1"
user_chip_file  <- "/localscratch/marisol/DASiRe/databases/encode_bedNarrowPeak_files/ENCFF520DIY.bed.gz" # This eventually will require modification, because the user is supposed to upload their own bed files.

```


## import user data and controls from encode

```{r}
# we want to use CHIP experiments of known splicing factors from encode to show as references in the plots.

# Import splicing factors list 
sf_list <- read.table(file = "/localscratch/marisol/DASiRe/databases/splicing_factors_list.txt",header = T,sep = "\t",fill = TRUE)

# import reference ENCODE datasets 
ENCODE_metadata <- read.table("/localscratch/marisol/DASiRe/databases/encode_bedNarrowPeak_files/metadata.tsv",sep="\t",header = TRUE)
ENCODE_metadata <- ENCODE_metadata %>%
  dplyr::filter(File.assembly=="GRCh38") %>%
  dplyr::filter(Output.type %in% c("optimal IDR thresholded peaks")) %>%
  dplyr::filter(Biosample.term.name=="K562") %>% # Maybe this can be also selected by the user
  dplyr::select(File.accession,Assay,Biosample.term.name,Experiment.target)
ENCODE_metadata$Experiment.target <- gsub(pattern = "-human",replacement = "",x = ENCODE_metadata$Experiment.target)
ENCODE_metadata <- ENCODE_metadata %>%
  mutate(Assay = if_else(Experiment.target %in% c(sf_list$Gene,"TARDBP","RBM25","RBFOX2","PTBP1","KHSRP","FUS","SF1","CELF1") | grepl(pattern = "HNRNP",x = Experiment.target),
                         true = "Splicing Factor",
                         false = Assay))

# These are the chip seq datasets of encode
input_chips <- ENCODE_metadata %>%
dplyr::filter(Assay == "Splicing Factor")

DT::datatable(input_chips)
```


```{r}
library(GenomicRanges)
library(rtracklayer)
library(IRanges)
library(genomation)
library(ChIPseeker)

all.chips <- c()
for (accession in 1:nrow(input_chips)) {
  all.chips[[input_chips$Experiment.target[accession]]] <- IRanges::reduce(readPeakFile(peakfile = paste0("/localscratch/marisol/DASiRe/databases/encode_bedNarrowPeak_files/",input_chips$File.accession[accession],".bed.gz")))
}

all.chips[[user_chip_name]] <- IRanges::reduce(readPeakFile(peakfile = user_chip_file),)

for (chip in 1:length(all.chips)){
  all.chips[[chip]]@seqnames <- gsub(pattern = "chr",replacement = "",x = all.chips[[chip]]@seqnames)
}
```



## Distribution of peaks in genomic regions
```{r}
library(EnsDb.Hsapiens.v86)
library(org.Hs.eg.db)
library(dplyr)
library(GenomicFeatures)
library(RMariaDB)
# txdb_object <- makeTxDbFromEnsembl(organism="Homo sapiens",release=NA,circ_seqs=NULL,server="ensembldb.ensembl.org",
#                                    username="anonymous", password=NULL, port=0L,tx_attrib=NULL)

txdb_object <- makeTxDbFromGFF(file = "/localscratch/marisol/DASiRe/databases/gencode_annotation_files/gencode.v36.annotation.gtf",format = "gtf")

library(ChIPpeakAnno)
peakAnno.edb <- c()
df_peakAnno.edb <- c()

for (i in 1:length(all.chips)){
peakAnno.edb[[names(all.chips)[i]]] <- assignChromosomeRegion(all.chips[[i]], nucleotideLevel=FALSE,
                           TxDb=txdb_object)
data_tmp <- as.data.frame(peakAnno.edb[[i]]$percentage) %>%  mutate(dataset=names(all.chips)[i])
df_peakAnno.edb <- as.data.frame(rbind(df_peakAnno.edb,data_tmp))
}

df_peakAnno.edb$subjectHits <- gsub(pattern = "fiveUTRs",replacement = "5'UTR",x = df_peakAnno.edb$subjectHits)
df_peakAnno.edb$subjectHits <- gsub(pattern = "threeUTRs",replacement = "3'UTR",x = df_peakAnno.edb$subjectHits)
df_peakAnno.edb$subjectHits <- gsub(pattern = "immediateDownstream",replacement = "Immediate downstream",x = df_peakAnno.edb$subjectHits)
df_peakAnno.edb$subjectHits <- gsub(pattern = "Intergenic.Region",replacement = "Intergenic",x = df_peakAnno.edb$subjectHits)


save(txdb_object, df_peakAnno.edb, file="peaksregion.RData")
library(ggplot2)
library(viridis)
renderPlot({
df_peakAnno.edb %>%
  ggplot( aes(fill=subjectHits, y=Freq, x=dataset)) +
  geom_bar(position="fill", stat="identity",color="black") +
  coord_flip()
})
```


## Visual evaluation of ChIP-seq data

```{r}

# Visualize chips

gene_to_display = "BTNL10" #This SHOULD be defined by the user, I selected this one as an example for this protein

# Get human genome reference track
library(biomaRt)
library(Gviz)
mart = useMart("ensembl",dataset="hsapiens_gene_ensembl")
fm = Gviz:::.getBMFeatureMap()
fm["symbol"] = "external_gene_name"

bm = BiomartGeneRegionTrack(biomart=mart,
                            size=2, name="GENCODE annotation", utr5="red", utr3="red",
                            protein_coding="black", col.line=NULL, #cex=7,
                            collapseTranscripts="longest",
                            featureMap=fm)
AT = GenomeAxisTrack( )

# create USER CHIP-seq track
TrackUserChip = AnnotationTrack(all.chips[[user_chip_name]], name=user_chip_name, shape='box',fill='blue',size=2,col="blue")

# Plot peaks in a gene that the user defines
mart_export <- read.table(file = "/localscratch/marisol/DASiRe/databases/mart_export.txt",sep = "\t",header = T)

renderPlot({
plotTracks(c(TrackUserChip, bm, AT),
           chromosome = mart_export$chromosome_name[mart_export$external_gene_name==gene_to_display],
           from=as.numeric(mart_export$start_position[mart_export$external_gene_name==gene_to_display][1]),
           to=as.numeric(mart_export$end_position[mart_export$external_gene_name==gene_to_display][1]),
           extend.right = 3000,extend.left = 3000,
           transcriptAnnotation="symbol",
           window="auto",
          cex.title=1, fontsize=8,
           type="histogram" )
})
```


# ChIP peak enrichment

Goal: Evaluate the association between the spliced genes and the presence of a protein 
compared to the non-spliced genes detected in a specific RNA-seq dataset (user RNA-seq input). 
The presence of a protein is taken from the ChiP-seq data that the user gives as input. 

We will include additional ChIP-seq data of known Splicing Factors available from ENCODE to 
use as reference for the user. 

The main steps are: 

1. Parse the results of the splicing analysis, for every splicing type detected by Majiq (ES, IR, A5SS, A3SS), DEXseq, and ISA:
  1.1. Create GenomicRanges object with the coordinates of the spliced genes
  1.2. Create GenomicRanges object with the coordinates of the non-spliced genes
  1.3. Create GenomicRanges object with the promoter coordinates of the spliced genes
  1.4. Create GenomicRanges object with the promoter coordinates of the non-spliced genes

3. Perform enrichment test by comparing the ChIP-seq peak frequencies in Spliced genes vs non spliced genes for the genes found by every tool

## 1. For every splicing type detected by Majiq (ES, IR, A5SS, A3SS), the genes detected by DEXseq and IsoformSwitchAnalyzer:

### MAJIQ results - Create GenomicRanges object with the coordinates of the spliced genes and the non-spliced genes

```{r}

# Import again the chip-seq data (for some reason, I cannot make the remaining code work with a specific annotation of the chromosomes)

all.chips <- c()
for (accession in 1:nrow(input_chips)) {
  all.chips[[input_chips$Experiment.target[accession]]] <- IRanges::reduce(readPeakFile(peakfile = paste0("/localscratch/marisol/DASiRe/databases/encode_bedNarrowPeak_files/",input_chips$File.accession[accession],".bed.gz")))
}
all.chips[[user_chip_name]] <- IRanges::reduce(readPeakFile(peakfile = user_chip_file),)

mart_export <- read.table(file = "/localscratch/marisol/DASiRe/databases/mart_export.txt",sep = "\t",header = T)
mart_export$chromosome_name <- paste0("chr",mart_export$chromosome_name)

################
#### ES ########
################

## Create granges object with GENES of POSITIVE ES events
# the final object is *majiq.gr.es.gene* which is the GR object with the genes that undergo exon skipping detected by majiq
pos_genes_to_select <- unique(majiq.es.pos$X.Gene.Name)
mart_filt <- mart_export %>%
  dplyr::filter(external_gene_name %in% pos_genes_to_select)
mart_filt <- unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])
majiq.gr.es.gene <-   GRanges(seqnames = Rle( mart_filt$chromosome_name),
                              ranges = IRanges( start = mart_filt$start_position,
                                                end = mart_filt$end_position),
                              strand = Rle( rep("*", nrow(mart_filt)) ),
                              gene = mart_filt$external_gene_name)

majiq.gr.es.promoter <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                          ranges = IRanges( start = mart_filt$start_position -200,
                                            end = mart_filt$start_position +200),
                          strand = Rle( rep("*", nrow(mart_filt)) ),
                          gene = mart_filt$external_gene_name)


## Create granges object with GENES of NEGATIVE ES events
# the final object is *majiq.gr.es.gene.neg* which is the GR object with the genes that undergo exon skipping detected by majiq

neg_genes_to_select <- unique(majiq.es.neg$X.Gene.Name)
mart_filt <- mart_export %>%
  dplyr::filter(external_gene_name %in% neg_genes_to_select)
mart_filt <-unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])
majiq.gr.es.gene.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                ranges = IRanges( start = mart_filt$start_position,
                                                  end = mart_filt$end_position),
                                strand = Rle( rep("*", nrow(mart_filt)) ),
                                gene = mart_filt$external_gene_name)

majiq.gr.es.promoter.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                    ranges = IRanges( start = mart_filt$start_position -200,
                                                      end = mart_filt$start_position +200),
                                    strand = Rle( rep("*", nrow(mart_filt)) ),
                                    gene = mart_filt$external_gene_name)

################
#### IR ########
################

## Create granges object with GENES of POSITIVE IR events
# the final object is *majiq.gr.ir.gene* which is the GR object with the genes that undergo intron retention detected by majiq

pos_genes_to_select <- unique(majiq.ir.pos$X.Gene.Name)
mart_filt <- mart_export %>%
  dplyr::filter(external_gene_name %in% pos_genes_to_select)
mart_filt <- unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])
majiq.gr.ir.gene <-   GRanges(seqnames = Rle( mart_filt$chromosome_name),
                              ranges = IRanges( start = mart_filt$start_position,
                                                end = mart_filt$end_position),
                              strand = Rle( rep("*", nrow(mart_filt)) ),
                              gene = mart_filt$external_gene_name)

majiq.gr.ir.promoter <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                          ranges = IRanges( start = mart_filt$start_position -200,
                                            end = mart_filt$start_position +200),
                          strand = Rle( rep("*", nrow(mart_filt)) ),
                          gene = mart_filt$external_gene_name)


## Create granges object with GENES of NEGATIVE IR events
# the final object is *majiq.gr.ir.gene.neg* which is the GR object with the genes that undergo intron retention detected by majiq

neg_genes_to_select <- unique(majiq.ir.neg$X.Gene.Name)
mart_filt <- mart_export %>%
  dplyr::filter(external_gene_name %in% neg_genes_to_select)
mart_filt <-unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])
majiq.gr.ir.gene.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                ranges = IRanges( start = mart_filt$start_position,
                                                  end = mart_filt$end_position),
                                strand = Rle( rep("*", nrow(mart_filt)) ),
                                gene = mart_filt$external_gene_name)

majiq.gr.ir.promoter.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                    ranges = IRanges( start = mart_filt$start_position -200,
                                                      end = mart_filt$start_position +200),
                                    strand = Rle( rep("*", nrow(mart_filt)) ),
                                    gene = mart_filt$external_gene_name)

################
#### A5SS ########
################

## Create granges object with GENES of POSITIVE A5SS events
# the final object is *majiq.gr.a5.gene* which is the GR object with the genes that undergo A5SS detected by majiq

pos_genes_to_select <- unique(majiq.a5.pos$X.Gene.Name)
mart_filt <- mart_export %>%
  dplyr::filter(external_gene_name %in% pos_genes_to_select)
mart_filt <- unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])
majiq.gr.a5.gene <-   GRanges(seqnames = Rle( mart_filt$chromosome_name),
                              ranges = IRanges( start = mart_filt$start_position,
                                                end = mart_filt$end_position),
                              strand = Rle( rep("*", nrow(mart_filt)) ),
                              gene = mart_filt$external_gene_name)

majiq.gr.a5.promoter <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                          ranges = IRanges( start = mart_filt$start_position -200,
                                            end = mart_filt$start_position +200),
                          strand = Rle( rep("*", nrow(mart_filt)) ),
                          gene = mart_filt$external_gene_name)


## Create granges object with GENES of NEGATIVE A5SS events
# the final object is *majiq.gr.a5.gene.neg* which is the GR object with the genes that undergo A5SS detected by majiq

neg_genes_to_select <- unique(majiq.a5.neg$X.Gene.Name)
mart_filt <- mart_export %>%
  dplyr::filter(external_gene_name %in% neg_genes_to_select)
mart_filt <-unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])
majiq.gr.a5.gene.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                ranges = IRanges( start = mart_filt$start_position,
                                                  end = mart_filt$end_position),
                                strand = Rle( rep("*", nrow(mart_filt)) ),
                                gene = mart_filt$external_gene_name)

majiq.gr.a5.promoter.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                    ranges = IRanges( start = mart_filt$start_position -200,
                                                      end = mart_filt$start_position +200),
                                    strand = Rle( rep("*", nrow(mart_filt)) ),
                                    gene = mart_filt$external_gene_name)

################
#### A3SS ########
################

## Create granges object with GENES of POSITIVE A3SS events
# the final object is *majiq.gr.a3.gene* which is the GR object with the genes that undergo A3SS detected by majiq

pos_genes_to_select <- unique(majiq.a3.pos$X.Gene.Name)
mart_filt <- mart_export %>%
  dplyr::filter(external_gene_name %in% pos_genes_to_select)
mart_filt <- unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])
majiq.gr.a3.gene <-   GRanges(seqnames = Rle( mart_filt$chromosome_name),
                              ranges = IRanges( start = mart_filt$start_position,
                                                end = mart_filt$end_position),
                              strand = Rle( rep("*", nrow(mart_filt)) ),
                              gene = mart_filt$external_gene_name)

majiq.gr.a3.promoter <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                          ranges = IRanges( start = mart_filt$start_position -200,
                                            end = mart_filt$start_position +200),
                          strand = Rle( rep("*", nrow(mart_filt)) ),
                          gene = mart_filt$external_gene_name)


## Create granges object with GENES of NEGATIVE A3SS events
# the final object is *majiq.gr.a3.gene.neg* which is the GR object with the genes that undergo A3SS detected by majiq

neg_genes_to_select <- unique(majiq.a3.neg$X.Gene.Name)
mart_filt <- mart_export %>%
  dplyr::filter(external_gene_name %in% neg_genes_to_select)
mart_filt <-unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])
majiq.gr.a3.gene.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                ranges = IRanges( start = mart_filt$start_position,
                                                  end = mart_filt$end_position),
                                strand = Rle( rep("*", nrow(mart_filt)) ),
                                gene = mart_filt$external_gene_name)

majiq.gr.a3.promoter.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                    ranges = IRanges( start = mart_filt$start_position -200,
                                                      end = mart_filt$start_position +200),
                                    strand = Rle( rep("*", nrow(mart_filt)) ),
                                    gene = mart_filt$external_gene_name)

rm(pos_genes_to_select,neg_genes_to_select,mart_filt)
```

### DEXseq results - Create GenomicRanges object with the coordinates of the spliced genes and the non-spliced genes

```{r}
################################################
#### Differential exon expression ########
################################################
load(file = "exon_counts/dxr1.RData")

DEXseq_res <- as.data.frame(dxr1)

## Create granges object with GENES of POSITIVE exons
# the final object is *dexseq.gr.ex.gene* which is the GR object with the genes with significant exon changes detected by DEXseq
pos_genes_to_select <- gsub(pattern = "\\..*",replacement = "",x = unique(DEXseq_res$groupID[DEXseq_res$padj<0.1]))
mart_filt <- mart_export %>%
  dplyr::filter(ensembl_gene_id %in% pos_genes_to_select)
mart_filt <- base::unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])
dexseq.gr.ex.gene <-   GRanges(seqnames = Rle( mart_filt$chromosome_name),
                              ranges = IRanges( start = mart_filt$start_position,
                                                end = mart_filt$end_position),
                              strand = Rle( rep("*", nrow(mart_filt)) ),
                              gene = mart_filt$external_gene_name)

dexseq.gr.ex.promoter <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                          ranges = IRanges( start = mart_filt$start_position -200,
                                            end = mart_filt$start_position +200),
                          strand = Rle( rep("*", nrow(mart_filt)) ),
                          gene = mart_filt$external_gene_name)

## Create granges object with GENES of NEGATIVE ES events
# the final object is *majiq.gr.es.gene.neg* which is the GR object with the genes that undergo exon skipping detected by majiq

neg_genes_to_select <- gsub(pattern = "\\..*",replacement = "",x = unique(DEXseq_res$groupID[DEXseq_res$padj>0.9]))
mart_filt <- mart_export %>%
  dplyr::filter(ensembl_gene_id %in% neg_genes_to_select)
mart_filt <-unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])
dexseq.gr.ex.gene.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                ranges = IRanges( start = mart_filt$start_position,
                                                  end = mart_filt$end_position),
                                strand = Rle( rep("*", nrow(mart_filt)) ),
                                gene = mart_filt$external_gene_name)

dexseq.gr.ex.promoter.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                    ranges = IRanges( start = mart_filt$start_position -200,
                                                      end = mart_filt$start_position +200),
                                    strand = Rle( rep("*", nrow(mart_filt)) ),
                                    gene = mart_filt$external_gene_name)

rm(pos_genes_to_select,neg_genes_to_select,mart_filt)
```

### ISA results - Create GenomicRanges object with the coordinates of the spliced genes and the non-spliced genes

```{r}
################################################
#### Genes with isoform switches ########
################################################

# Parse IsoformSwitchAnalyzer output
load(file = "pseudocounts/exampleSwitchListAnalyzed.RData")
isa_res_pos <- IsoformSwitchAnalyzeR::extractTopSwitches(exampleSwitchListAnalyzed,alpha = 0.05,dIFcutoff = 0.1,n=Inf)
isa_res_neg <- IsoformSwitchAnalyzeR::extractTopSwitches(exampleSwitchListAnalyzed,alpha = 1,dIFcutoff = 0,n=Inf)
isa_res.temp <- IsoformSwitchAnalyzeR::extractTopSwitches(exampleSwitchListAnalyzed,alpha = 0.9,dIFcutoff = 0,n=Inf)
isa_res_neg <- isa_res_neg %>%
  dplyr::filter(gene_name %notin% isa_res.temp$gene_name)
rm(isa_res.temp)

## Create granges object with GENES and PROMOTERS of POSITIVE isoforms
# the final object is *isa.gr.iso.gene* which is the GR object with the genes with significant exon changes detected by DEXseq
pos_genes_to_select <- isa_res_pos$gene_name
mart_filt <- mart_export %>%
  dplyr::filter(external_gene_name %in% pos_genes_to_select)
mart_filt <- unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])

isa.gr.iso.gene <-   GRanges(seqnames = Rle( mart_filt$chromosome_name),
                              ranges = IRanges( start = mart_filt$start_position,
                                                end = mart_filt$end_position),
                              strand = Rle( rep("*", nrow(mart_filt)) ),
                              gene = mart_filt$external_gene_name)

isa.gr.iso.promoter <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                          ranges = IRanges( start = mart_filt$start_position -200,
                                            end = mart_filt$start_position +200),
                          strand = Rle( rep("*", nrow(mart_filt)) ),
                          gene = mart_filt$external_gene_name)

## Create granges object with GENES and PROMOTERS of NEGATIVE isoforms
# the final object is *majiq.gr.es.gene.neg* which is the GR object with the genes that undergo exon skipping detected by majiq

neg_genes_to_select <- isa_res_neg$gene_name
mart_filt <- mart_export %>%
  dplyr::filter(external_gene_name %in% neg_genes_to_select)
mart_filt <-unique(mart_filt[c("ensembl_gene_id","external_gene_name","start_position","end_position","chromosome_name")])

isa.gr.iso.gene.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                ranges = IRanges( start = mart_filt$start_position,
                                                  end = mart_filt$end_position),
                                strand = Rle( rep("*", nrow(mart_filt)) ),
                                gene = mart_filt$external_gene_name)

isa.gr.iso.promoter.neg <- GRanges(seqnames = Rle( mart_filt$chromosome_name),
                                    ranges = IRanges( start = mart_filt$start_position -200,
                                                      end = mart_filt$start_position +200),
                                    strand = Rle( rep("*", nrow(mart_filt)) ),
                                    gene = mart_filt$external_gene_name)

rm(pos_genes_to_select,neg_genes_to_select,mart_filt)
```


## 2. Perform enrichment test by comparing the spliced and non-spliced genes frequencies with and whithout chip-seq peaks

Steps
For each splicing event type and for each protein with ChIP-seq data:
  2.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
  2.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
  2.3. Generate the contingency table
  2.4. Perform statistical test (Chi square)
  2.5. Plot results

### GENE level enrichment
```{r}
################
#### MAJIQ- IR ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(majiq.gr.ir.gene,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(majiq.gr.ir.gene.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(majiq.gr.ir.gene))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(majiq.gr.ir.gene.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
enrichment_results <- c() 
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results <- as.data.frame(rbind(enrichment_results,
                                 c(pval,estimate,chip,"Majiq - Intron Retention")))
}

################
#### MAJIQ- ES ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(majiq.gr.es.gene,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(majiq.gr.es.gene.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(majiq.gr.es.gene))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(majiq.gr.es.gene.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results <- as.data.frame(rbind(enrichment_results,
                                 c(pval,estimate,chip,"Majiq - Exon Skipping")))
}


################
#### MAJIQ- A5SS ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(majiq.gr.a5.gene,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(majiq.gr.a5.gene.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(majiq.gr.a5.gene))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(majiq.gr.a5.gene.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results <- as.data.frame(rbind(enrichment_results,
                                 c(pval,estimate,chip,"Majiq - A5SS")))
}

################
#### MAJIQ- A3SS ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(majiq.gr.a3.gene,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(majiq.gr.a3.gene.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(majiq.gr.a3.gene))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(majiq.gr.a3.gene.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results <- as.data.frame(rbind(enrichment_results,
                                 c(pval,estimate,chip,"Majiq - A3SS")))
}


################
#### DEXseq - Exons ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(dexseq.gr.ex.gene,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(dexseq.gr.ex.gene.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(dexseq.gr.ex.gene))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(dexseq.gr.ex.gene.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results <- as.data.frame(rbind(enrichment_results,
                                 c(pval,estimate,chip,"DEXseq - Differential exon")))
}


################
#### IsoformSwitchAnalyzer - Isoform ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(isa.gr.iso.gene,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(isa.gr.iso.gene.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(isa.gr.iso.gene))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(isa.gr.iso.gene.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results <- as.data.frame(rbind(enrichment_results,
                                 c(pval,estimate,chip,"IsoformSwitchAnalyzer - Isoform switch")))
}

colnames(enrichment_results) <- c("pvalue","estimate","chip_dataset","Event type - Tool")

#3.5 Plot results
library(ggplot2)
ht_data <- enrichment_results
ht_data$pvalue[as.numeric(ht_data$pvalue) > 0.05] <- NA
ht_data$pvalue[as.numeric(ht_data$estimate) == "Inf"] <- NA


ggplot(ht_data, aes(x = chip_dataset, y = `Event type - Tool`, fill = as.numeric(pvalue))) + 
  geom_tile() + 
  labs(fill = "Chi Square pvalue") + 
  ggtitle("Enrichment analysis of ChIP-seq\npeak in spliced genes")+
  geom_text(aes(label=round(as.numeric(estimate),digits = 2))) +
  scale_fill_gradientn(colors= rev(RColorBrewer::brewer.pal(n = 6,name = "BuPu")),na.value = "white")

```


### PROMOTER level enrichment
```{r}
################
#### MAJIQ- IR ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(majiq.gr.ir.promoter,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(majiq.gr.ir.promoter.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(majiq.gr.ir.promoter))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(majiq.gr.ir.promoter.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
enrichment_results_promoter <- c() 
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results_promoter <- as.data.frame(rbind(enrichment_results_promoter,
                                 c(pval,estimate,chip,"Majiq - Intron Retention")))
}

################
#### MAJIQ- ES ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(majiq.gr.es.promoter,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(majiq.gr.es.promoter.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(majiq.gr.es.promoter))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(majiq.gr.es.promoter.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results_promoter <- as.data.frame(rbind(enrichment_results_promoter,
                                 c(pval,estimate,chip,"Majiq - Exon Skipping")))
}


################
#### MAJIQ- A5SS ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(majiq.gr.a5.promoter,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(majiq.gr.a5.promoter.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(majiq.gr.a5.promoter))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(majiq.gr.a5.promoter.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results_promoter <- as.data.frame(rbind(enrichment_results_promoter,
                                 c(pval,estimate,chip,"Majiq - A5SS")))
}

################
#### MAJIQ- A3SS ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(majiq.gr.a3.promoter,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(majiq.gr.a3.promoter.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(majiq.gr.a3.promoter))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(majiq.gr.a3.promoter.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results_promoter <- as.data.frame(rbind(enrichment_results_promoter,
                                 c(pval,estimate,chip,"Majiq - A3SS")))
}


################
#### DEXseq - Exons ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(dexseq.gr.ex.promoter,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(dexseq.gr.ex.promoter.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(dexseq.gr.ex.promoter))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(dexseq.gr.ex.promoter.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results_promoter <- as.data.frame(rbind(enrichment_results_promoter,
                                 c(pval,estimate,chip,"DEXseq - Differential exon")))
}


################
#### IsoformSwitchAnalyzer - Isoform ########
################
#3.1. Calculate the frequency of POSITIVE spliced genes with at least one chip-seq peak and POSITIVE spliced genes without chip-seq peak
#3.2. Calculate the frequency of NEGATIVE spliced genes with at least one chip-seq peak and NEGATIVE spliced genes without chip-seq peak
library(dplyr)
library(GenomicRanges)
obsFreq_interest <- c()
for (chip in 1:length(all.chips)){
  # freq_bg=length(subsetByOverlaps(gr.gencode.genes,all_chips[[chip]]))
  freq_as=length(subsetByOverlaps(isa.gr.iso.promoter,all.chips[[chip]]))
  freq_nas=length(subsetByOverlaps(isa.gr.iso.promoter.neg,all.chips[[chip]]))
  obsFreq_interest <- as.data.frame(rbind(obsFreq_interest,
                               c(as.numeric(freq_as),as.numeric(length(isa.gr.iso.promoter))-as.numeric(freq_as),
                                 as.numeric(freq_nas),as.numeric(length(isa.gr.iso.promoter.neg))-as.numeric(freq_nas),
                                 # as.numeric(freq_bg),as.numeric(length(gr.gencode.genes))-as.numeric(freq_bg),
                                 names(all.chips)[chip])))
}
colnames(obsFreq_interest) <- c("freqASwithPeak","freqASnoPeak",
                       "freqNASwithPeak","freqNASnoPeak",
                       # "freqBGwithPeak","freqBGnoPeak",
                       "chip_protein")

#3.3. Generate the contingency table
library(reshape)
final_contingency_table <- c()
contingency_table <- melt(data = obsFreq_interest,id.vars = "chip_protein")
contingency_table <- contingency_table %>%
  dplyr::mutate(event=if_else(condition = grepl(pattern = "NAS",x = contingency_table$variable),true = "NAS","AS"))  %>%
  dplyr::mutate(peak=if_else(condition = grepl(pattern = "withPeak",x = contingency_table$variable),true = "Peak","noPeak"))
contingency_table$variable <- NULL

for (chip in names(all.chips)){
  df <- reshape::cast(contingency_table,formula = peak ~event,value.var = "value",fun.aggregate = NULL,subset = chip_protein== chip)
  df <- df %>% 
    dplyr::mutate(chip_protein = chip)
  df <- df[c(2,1),]
  final_contingency_table <- as.data.frame(rbind(final_contingency_table,
                                                 df))
  }
rm(df)

#3.4. Perform statistical test (Chi square)
for (chip in names(all.chips)){
  chip_df <- final_contingency_table %>%
    dplyr::filter(chip_protein == chip)%>%
    dplyr::select(AS,NAS)
  chip_df <- apply(X = chip_df,MARGIN = 2,FUN = as.numeric)
  rownames(chip_df) <- c("noPeak","Peak")
  chisq <- fisher.test(chip_df,alternative = "two.sided")
  pval <- chisq$p.value
  estimate <- chisq$estimate
  enrichment_results_promoter <- as.data.frame(rbind(enrichment_results_promoter,
                                 c(pval,estimate,chip,"IsoformSwitchAnalyzer - Isoform switch")))
}

colnames(enrichment_results_promoter) <- c("pvalue","estimate","chip_dataset","Event type - Tool")

#3.5 Plot results
library(ggplot2)
ht_data_promoter <- enrichment_results_promoter
ht_data_promoter$pvalue[as.numeric(ht_data_promoter$pvalue) > 0.05] <- NA
ht_data_promoter$pvalue[as.numeric(ht_data_promoter$estimate) == "Inf"] <- NA

renderPlot({
ggplot(ht_data_promoter, aes(x = chip_dataset, y = `Event type - Tool`, fill = as.numeric(pvalue))) + 
  geom_tile() + 
  labs(fill = "Chi Square\npvalue") + 
  ggtitle("Enrichment analysis of ChIP-seq\npeak in promoters of spliced genes")+
  geom_text(aes(label=round(as.numeric(estimate),digits = 2))) +
  scale_fill_gradientn(colors= rev(RColorBrewer::brewer.pal(n = 6,name = "BuPu")),na.value = "white")
})
```